In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data= pd.read_csv("churn_telco.csv")

In [4]:
data.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [13]:
# Data Overview
print("Rows",data.shape[0])
print("Columns",data.shape[1])
print("Missing values",data.isna().sum().values.sum())
print("Unique Values",data.nunique())

Rows 7043
Columns 21
Missing values 0
Unique Values customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64


tenure has 73 unique numbers distributed among 7043 customer Id's. 
Converting tenure from monthly to yearly makes more sense. it will help in building a good model as well. 

In [44]:
#replacing empty values in total charges to null
data["TotalCharges"]= data["TotalCharges"].replace(" ",np.nan)

data= data[data["TotalCharges"].notnull()]
data=data.reset_index()[data.columns]

data["TotalCharges"]= data["TotalCharges"].astype(float)

#replacing no internet service on some columns to just no
cols_no_internet = ["OnlineSecurity","OnlineBackup","DeviceProtection","TechSupport","StreamingTV","StreamingMovies"]

for cols in cols_no_internet:
    data[cols]=data[cols].replace({'No internet service':'No'})
    
#senior citizens replace values
data["SeniorCitizen"]=data["SeniorCitizen"].replace({1:'Yes',0:'No'})

# converting monthly tenure to yearly tenure

def tenure_convert(data):
    if (data['tenure']>0) & (data['tenure']<=12):
        return 'tenure_Year1'
    elif (data['tenure']>12) & (data['tenure']<=24):
        return 'tenure_Year2'
    elif (data['tenure']>24) & (data['tenure'] <=36):
        return 'tenure_Year3'
    elif (data['tenure']>36) & (data['tenure'] <= 48):
        return 'tenure_Year4'
    elif (data['tenure']>48) & (data['tenure'] <= 60):
        return 'tenure_Year5'
    elif (data['tenure']>60) & (data['tenure'] <= 72):
        return 'tenure_Year6'
    
data["tenure_year"]= data.apply(lambda data:tenure_convert(data),axis=1)
    

In [32]:
data=data.reset_index()[data.columns]

In [39]:
data.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [45]:
data

customerID  gender SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female            No     Yes         No       1   
1     5575-GNVDE    Male            No      No         No      34   
2     3668-QPYBK    Male            No      No         No       2   
3     7795-CFOCW    Male            No      No         No      45   
4     9237-HQITU  Female            No      No         No       2   
...          ...     ...           ...     ...        ...     ...   
7027  6840-RESVB    Male            No     Yes        Yes      24   
7028  2234-XADUH  Female            No     Yes        Yes      72   
7029  4801-JZAZL  Female            No     Yes        Yes      11   
7030  8361-LTMKD    Male           Yes     Yes         No       4   
7031  3186-AJIEK    Male            No      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7027          Yes               Yes             DSL            Yes  ...   
7028          Yes               Yes     Fiber optic             No  ...   
7029           No  No phone service             DSL            Yes  ...   
7030          Yes               Yes     Fiber optic             No  ...   
7031          Yes                No     Fiber optic            Yes  ...   

     TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0             No          No              No  Month-to-month              Yes   
1             No          No              No        One year               No   
2             No          No              No  Month-to-month              Yes   
3            Yes          No              No        One year               No   
4             No          No              No  Month-to-month              Yes   
...          ...         ...             ...             ...              ...   
7027         Yes         Yes             Yes        One year              Yes   
7028          No         Yes             Yes        One year              Yes   
7029          No          No              No  Month-to-month              Yes   
7030          No          No              No  Month-to-month              Yes   
7031         Yes         Yes             Yes        Two year              Yes   

                  PaymentMethod MonthlyCharges TotalCharges  Churn  \
0              Electronic check          29.85        29.85     No   
1                  Mailed check          56.95      1889.50     No   
2                  Mailed check          53.85       108.15    Yes   
3     Bank transfer (automatic)          42.30      1840.75     No   
4              Electronic check          70.70       151.65    Yes   
...                         ...            ...          ...    ...   
7027               Mailed check          84.80      1990.50     No   
7028    Credit card (automatic)         103.20      7362.90     No   
7029           Electronic check          29.60       346.45     No   
7030               Mailed check          74.40       306.60    Yes   
7031  Bank transfer (automatic)         105.65      6844.50     No   

       tenure_year  
0     tenure_Year1  
1     tenure_Year3  
2     tenure_Year1  
3     tenure_Year4  
4     tenure_Year1  
...            ...  
7027  tenure_Year2  
7028  tenure_Year6  
7029  tenure_Year1  
7030  tenure_Year1  
7031  tenure_Year6  

[7032 rows x 22 columns]

In [ ]:
#Scaling and Encoding